avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java

```diff
@@ -71,4 +78,28 @@ protected Object deserialize(byte[] payload) throws SerializationException {
       throw new SerializationException("Error deserializing Avro message for id " + id, e);
     }
   }
+
+    private DatumReader getDatumReader(Schema schema) {
+        if (useSpecificAvroReader) {
+            return new SpecificDatumReader(schema, getTargetSchema(schema));
+        } else {
+            return new GenericDatumReader(schema);
+        }
+    }
+
+    private Schema getTargetSchema(Schema schema) {
+        Schema targetSchema = targetSchemaCache.get(schema);
```

Comment:
This cache needs synchronization -- this may be accessed by multiple threads. A simple concurrent map should work fine.